File: backend/app/vibe/path_security.py
"""Secure path validation for worker project roots."""

from __future__ import annotations

import os
from pathlib import Path


def normalize_path(path: str) -> Path:
    return Path(os.path.normpath(os.path.abspath(path)))


def is_path_within_root(candidate: str, root: str) -> bool:
    """Return True if candidate is inside root (prevents directory traversal)."""
    try:
        cand = normalize_path(candidate)
        root_path = normalize_path(root)
        cand.relative_to(root_path)
        return True
    except (ValueError, OSError):
        return False


def resolve_project_path(project_path: str, allowed_roots: list[str]) -> str | None:
    """Resolve project path if it lies within any allowed root."""
    normalized = normalize_path(project_path)
    for root in allowed_roots:
        if is_path_within_root(str(normalized), root):
            return str(normalized)
    return None


def discover_projects(roots: list[str]) -> list[dict[str, str]]:
    """Discover git projects under configured roots."""
    projects: list[dict[str, str]] = []
    seen: set[str] = set()
    for root in roots:
        root_path = normalize_path(root)
        if not root_path.is_dir():
            continue
        candidates = [root_path]
        try:
            for child in root_path.iterdir():
                if child.is_dir() and not child.name.startswith("."):
                    candidates.append(child)
        except OSError:
            continue
        for path in candidates:
            key = str(path)
            if key in seen:
                continue
            seen.add(key)
            is_git = (path / ".git").exists()
            projects.append(
                {
                    "name": path.name,
                    "local_path": key,
                    "is_git": str(is_git).lower(),
                }
            )
    return projects

File: workers/vibe-coding-worker/src/path-security.ts
import fs from "node:fs";
import path from "node:path";

export function normalizePath(p: string): string {
  return path.resolve(path.normalize(p));
}

export function isPathWithinRoot(candidate: string, root: string): boolean {
  const cand = normalizePath(candidate);
  const rootPath = normalizePath(root);
  const rel = path.relative(rootPath, cand);
  return rel !== ".." && !rel.startsWith(`..${path.sep}`) && !path.isAbsolute(rel);
}

export function resolveProjectPath(projectPath: string, allowedRoots: string[]): string | null {
  const normalized = normalizePath(projectPath);
  for (const root of allowedRoots) {
    if (isPathWithinRoot(normalized, root)) return normalized;
  }
  return null;
}

export function discoverProjects(roots: string[]): Array<{ name: string; local_path: string; default_branch: string }> {
  const projects: Array<{ name: string; local_path: string; default_branch: string }> = [];
  const seen = new Set<string>();
  for (const root of roots) {
    const rootPath = normalizePath(root);
    if (!fs.existsSync(rootPath) || !fs.statSync(rootPath).isDirectory()) continue;
    const candidates = [rootPath];
    try {
      for (const entry of fs.readdirSync(rootPath, { withFileTypes: true })) {
        if (entry.isDirectory() && !entry.name.startsWith(".")) {
          candidates.push(path.join(rootPath, entry.name));
        }
      }
    } catch {
      continue;
    }
    for (const p of candidates) {
      if (seen.has(p)) continue;
      seen.add(p);
      projects.push({
        name: path.basename(p),
        local_path: p,
        default_branch: "main",
      });
    }
  }
  return projects;
}
